geonode/geoserver/views.py

```diff
@@ -477,21 +478,16 @@ def layer_acls(request):
                                 status=401,
                                 mimetype="text/plain")
 
-    # Include permissions on the anonymous user
-    all_readable = get_objects_for_user(acl_user, 'base.view_resourcebase')
-    all_writable_layers = get_objects_for_user(acl_user, 'layers.change_layer_data')
-    all_writable = []
-    for obj in all_writable_layers:
-        all_writable.append(obj.get_self_resource())
-
-    read_only = [
-        x.layer.typename for x in all_readable if x not in all_writable and hasattr(
-            x,
-            'layer')]
-    read_write = [
-        x.layer.typename for x in all_writable if x in all_readable and hasattr(
-            x,
-            'layer')]
+    # Include permissions on the anonymous user    
+    # use of polymorphic selectors/functions to optimize performances
+    layer_readable = get_objects_for_user(acl_user, 'view_resourcebase', ResourceBase.objects.instance_of(Layer)) 
+    layer_writable = get_objects_for_user(acl_user, 'change_layer_data', Layer.objects.all())
+
+    _read  = [l.typename for l in layer_readable.get_real_instances()]
+    _write = layer_writable.values_list('typename', flat=True)
```

Comment:
For the size of the `_read` and `_write` lists, it probably doesn't make a huge performance difference (especially when compared to the gain made by reducing DB queries) but if these were `set` instead of `list`, the operations below would be faster and could expressed simpler:

```python
read_only = read ^ write
write_only = read & write
```